neo/Consensus/ConsensusService.cs

```diff
@@ -225,9 +286,281 @@ private void OnPersistCompleted(Block block)
+                if (message.PrepareWitnessInvocationScripts[i] == null) continue;
+                var regeneratedPrepareResponse = tempContext.RegenerateSignedPayload(prepareResponseMsg, (ushort) i,
+                    message.PrepareWitnessInvocationScripts[i], message.PrepareTimestamps[i]);
+                if (!regeneratedPrepareResponse.Verify(snap) || !PerformBasicConsensusPayloadPreChecks(regeneratedPrepareResponse)) continue;
+                prepareResponses.Add((regeneratedPrepareResponse, prepareResponseMsg));
+                // Verify that there are M valid preparations, 1 Prepare Request + (M-1) Prepare responses
+                if (prepareRequest == null || prepareResponses.Count < context.M - 1) continue;
+                canRestoreView = true;
+                break;
+            }
+
+            // Only accept recovery from lower views if there were at least M valid prepare requests
+            if (message.ViewNumber < context.ViewNumber && !canRestoreView) return;
+
+            var verifiedChangeViewWitnessInvocationScripts = new byte[context.Validators.Length][];
+            if (!canRestoreView && message.ChangeViewWitnessInvocationScripts != null)
+            {
+                if (message.ChangeViewWitnessInvocationScripts.Length < context.Validators.Length) return;
+                if (message.ChangeViewTimestamps.Length < context.Validators.Length) return;
+                if (message.OriginalChangeViewNumbers.Length < context.Validators.Length) return;
+
+                var changeViewMsg = new ChangeView { NewViewNumber = message.ViewNumber };
+                int validChangeViewCount = 0;
+                for (int i = 0; i < context.Validators.Length; i++)
+                {
+                    if (message.ChangeViewWitnessInvocationScripts[i] == null) continue;
+
+                    changeViewMsg.ViewNumber = message.OriginalChangeViewNumbers[i];
+                    // Regenerate the ChangeView message
+                    var regeneratedChangeView = tempContext.RegenerateSignedPayload(changeViewMsg, (ushort) i,
+                        message.ChangeViewWitnessInvocationScripts[i], message.ChangeViewTimestamps[i]);
+                    if (!regeneratedChangeView.Verify(snap) || !PerformBasicConsensusPayloadPreChecks(regeneratedChangeView)) continue;
+                    verifiedChangeViewWitnessInvocationScripts[i] = message.ChangeViewWitnessInvocationScripts[i];
+                    validChangeViewCount++;
+                    if (validChangeViewCount < context.M) continue;
+                    canRestoreView = true;
+                    break;
+                }
+            }
+
+            if (!canRestoreView) return;
+            // We had enough valid change view messages or preparations to safely change the view number.
+            Log($"regenerating view: {message.ViewNumber}");
+
+            if (context.PrimaryIndex == context.MyIndex && prepareRequestPayload != null)
```

Comment:
But at this point was the correct View already set in the context?
Maybe it need an adjust in the check considering the Primary values on that view that will be set (`message.ViewNumber`).

Reply:
This was a nice catch. You are correct. I'm adjusting it to fix it.

Reply:
Just pushed a fix. It happens to be that when I tested restoring the primary that I just happened to be coming from a view where the primary was the same tempContext.PrimaryIndex. With further testing this would have been a problem.